File: containerWithWater.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach
/*
 * Using the 2 pointer approach, calculate the area and assign max with maximum area. 
 * Move the pointer based on the height to find the maximum area.
 */
public class containerWithWater {

    public int maxArea(int[] height) {
        int max = 0;
        int left = 0, right = height.length - 1;

        while (left < right) {
            int currArea = Math.min(height[left], height[right]) * (right - left);
            max = Math.max(max, currArea);
            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }

        return max;
    }

}

File: containerWithWater.py
'''
Time Complexity --> O(n)
Space Complexity --> O(1)
'''

class Solution:
    def maxArea(self, height: List[int]) -> int:
        l = 0
        h = len(height)-1
        max_val = 0
        while l<h:
            max_val = max(max_val,min(height[l],height[h])*(h-l))
            if height[l]<height[h]:
                l+=1
            elif height[l]>height[h]:
                h-=1
            else:
                l+=1
                h-=1
        return(max_val)
            
        